src/core/OSTreeTUI.hpp

```diff
@@ -3,20 +3,139 @@
+    void refresh_commitComponents();
+
+    /// @brief Refresh Level 2: Refreshes the commit list component & upper levels
+    void refresh_commitListComoponent();
+
+    /// @brief Refresh Level 1: Refreshes complete repository & upper levels
+    bool refresh_repository();
 
     /**
-     * @brief OSTree TUI main
+     * @brief sets the promotion mode
+     * @param active activate (true), or deactivate (false) promotion mode
+     * @param hash provide if setting mode to true
+     * @return false, if other promotion gets overwritten
+     */
+    bool setPromotionMode(bool active, std::string hash = "", bool setPromotionBranch = true);
+
+    /** @brief promote a commit
+     * @param hash 
+     * @param branch 
+     * @return promotion success
+     */ 
+    bool promoteCommit(std::string hash, std::string branch, std::vector<std::string> metadataStrings = {}, std::string newSubject = "", bool keepMetadata = true);
+
+private:
+    /**
+     * @brief Calculates all visible commits from an OSTreeRepo and a list of branches.
      * 
-     * @param repo ostree repository path
+     * @param repo OSTreeRepo
+     * @param visibleBranches Map: branch name -> visible
+     * @return Complete list of commit hashes in repo, that are part of the given branches 
+     */
+    std::vector<std::string> parseVisibleCommitMap(cpplibostree::OSTreeRepo& repo, std::unordered_map<std::string, bool>& visibleBranches);
+
+    /**
+     * @brief Adjust scroll offset to fit the selected commit
      */
-    int main(const std::string& repo, const std::vector<std::string>& startupBranches = {}, bool showTooltips = true);
+    void adjustScrollToSelectedCommit();
+
+public:
+    // SETTER
+    void setPromotionBranch(std::string promotionBranch);
+    void setSelectedCommit(size_t selectedCommit);
+
+    // non-const GETTER
+    std::vector<std::string>& getColumnToBranchMap();
+    ftxui::ScreenInteractive& getScreen();
+    
+    // GETTER
+    const cpplibostree::OSTreeRepo& getOstreeRepo() const;
+    const size_t& getSelectedCommit() const;
+    const std::string& getPromotionBranch() const;
+    const std::unordered_map<std::string, bool>& getVisibleBranches() const;
+    const std::vector<std::string>& getColumnToBranchMap() const;
+    const std::vector<std::string>& getVisibleCommitViewMap() const;
+    const std::unordered_map<std::string, ftxui::Color>& getBranchColorMap() const;
+    const int& getScrollOffset() const;
+    const bool& getInPromotionSelection() const;
+    const std::string& getPromotionHash() const;
 
+private:
+    // model
+    cpplibostree::OSTreeRepo ostreeRepo;
+
+    // backend states
+    size_t selectedCommit;
+    std::unordered_map<std::string, bool>  visibleBranches;         // map branch -> visibe
+	std::vector<std::string> columnToBranchMap;                     // map branch -> column in commit-tree (may be merged into one data-structure with visibleBranches)
```

Comment:
I will have to take a look into the CI there. It is interesting, because I thought I had tested that and also had states, where it failed, but surprisingly it didn't fail here. I'll create an issue for that.